internal/receiver/discoveryreceiver/endpoint_tracker.go

```diff
@@ -0,0 +1,203 @@
+// Copyright Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discoveryreceiver
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type endpointState string
+
+const (
+	addedState   endpointState = "added"
+	removedState endpointState = "removed"
+	changedState endpointState = "changed"
+)
+
+var (
+	_ observer.Notify = (*notify)(nil)
+)
+
+type endpointTracker struct {
+	logger       *zap.Logger
+	pLogs        chan plog.Logs
+	observables  map[config.ComponentID]observer.Observable
+	notifies     []*notify
+	logEndpoints bool
+}
+
+type notify struct {
+	observable      observer.Observable
+	endpointTracker *endpointTracker
+	observerID      config.ComponentID
+	id              observer.NotifyID
+}
+
+func newEndpointTracker(observables map[config.ComponentID]observer.Observable, config *Config, logger *zap.Logger, pLogs chan plog.Logs) *endpointTracker {
+	return &endpointTracker{
+		logEndpoints: config.LogEndpoints,
+		observables:  observables,
+		logger:       logger,
+		pLogs:        pLogs,
+	}
+}
+
+func (el *endpointTracker) start() {
```

Comment:
Nit: Should el be et?

Reply:
This was originally an "endpoint logger" but it has another responsibility to be added later so I renamed.